runtime/src/iree/hal/drivers/hip/event_semaphore.c

```diff
@@ -99,484 +132,749 @@ static void iree_hal_hip_semaphore_destroy(
       iree_hal_hip_semaphore_cast(base_semaphore);
   iree_allocator_t host_allocator = semaphore->host_allocator;
   IREE_TRACE_ZONE_BEGIN(z0);
-  IREE_IGNORE_ERROR(
-      iree_hal_hip_set_context(semaphore->symbols, semaphore->hip_context));
 
   iree_status_ignore(semaphore->failure_status);
   iree_slim_mutex_deinitialize(&semaphore->mutex);
 
-  iree_hal_semaphore_deinitialize(&semaphore->base);
+  iree_notification_deinitialize(&semaphore->state_notification);
+  for (iree_hal_hip_util_tree_node_t* i =
+           iree_hal_hip_util_tree_first(&semaphore->event_queue.tree);
+       i != NULL; i = iree_hal_hip_util_tree_node_next(i)) {
+    iree_hal_hip_event_t* event = ((iree_hal_hip_semaphore_queue_item_t*)
+                                       iree_hal_hip_util_tree_node_get_value(i))
+                                      ->event;
+    if (event) {
+      iree_hal_hip_event_release(event);
+    }
```

Comment:
release handles NULL events so you can just pass it directly
```suggestion
    iree_hal_hip_event_release(event);
```